doc/devguide/extending/app-layer/transactions.rst

```diff
@@ -0,0 +1,310 @@
+
+This is controlled by implementing progress states. In Suricata, those will be enums that are incremented as the parsing
+progresses. A state will start at 0. The higher its value, the closer the transaction would be to completion.
+
+The engine interacts with transactions state using a set of callbacks the parser registers. State is defined per flow direction (``STREAM_TOSERVER`` / ``STREAM_TOCLIENT``).
+
+In Summary - Transactions and State
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Initial state value: ``0``.
+- Simpler scenarios: state is simply an int.  ``1`` represents transaction completion, per direction.
+- Complex Transaction State in Suricata: ``enum`` (Rust: ``i32``). Completion is indicated by the highest enum value (some examples are: SSH, HTTP, HTTP2, DNS, SMB).
+
+Examples
+========
+
+This section shares some examples from Suricata codebase, to help visualize how Transactions work and are handled by the engine.
+
+Enums
+~~~~~
+
+Code snippet from: rust/src/ssh/ssh.rs:
+
+.. code-block:: rust
+
+    pub enum SSHConnectionState {
+        SshStateInProgress = 0,
+        SshStateBannerWaitEol = 1,
+        SshStateBannerDone = 2,
+        SshStateFinished = 3,
+    }
+
+From src/app-layer-ftp.h:
+
+.. code-block:: c
+
+    enum {
+        FTP_STATE_IN_PROGRESS,
+        FTP_STATE_PORT_DONE,
+        FTP_STATE_FINISHED,
+    };
+
+API Callbacks
+~~~~~~~~~~~~~
+
+In Rust, this is done via the RustParser struct. As seen in rust/src/applayer.rs:
+
+.. code-block:: rust
+
+    /// Rust parser declaration
+    pub struct RustParser {
+            .
+            .
+            .
+        /// Progress values at which the tx is considered complete in a direction
+        pub tx_comp_st_ts:      c_int,
+        pub tx_comp_st_tc:      c_int,
+        .
+        .
+        .
+    }
+
+In C, the callback API is:
+
+.. code-block:: c
+
+    void AppLayerParserRegisterStateProgressCompletionStatus(
+        AppProto alproto, const int ts, const int tc)
+
+Simple scenario described, in Rust:
+
+rust/src/dhcp/dhcp.rs:
+
+.. code-block:: rust
+
+    tx_comp_st_ts: 1
+    tx_comp_st_tc: 1
+
+For SSH, this looks like this:
+
+rust/src/ssh/ssh.rs:
+
+.. code-block:: rust
+
+    tx_comp_st_ts: SSHConnectionState::SshStateFinished as i32,
+    tx_comp_st_tc: SSHConnectionState::SshStateFinished as i32,
+
+In C, callback usage would be as follows:
+
+src/app-layer-dcerpc.c:
+
+.. code-block:: c
+
+    AppLayerParserRegisterStateProgressCompletionStatus(ALPROTO_DCERPC, 1, 1);
+
+src/app-layer-ftp.c:
+
+.. code-block:: c
+
+    AppLayerParserRegisterStateProgressCompletionStatus(
+        ALPROTO_FTP, FTP_STATE_FINISHED, FTP_STATE_FINISHED);
+
+Sequence Diagrams
+~~~~~~~~~~~~~~~~~
+
+A DNS transaction in Suricata can be considered unidirectional:
```

Comment:
The image here shows that both request and response are part of the transaction, which would make it bidirectional. The implementation for dns is unidirectional, which means that both request and response get their own tx.

Reply:
Hm, I updated this following a comment from @jasonish in my previous pr (https://github.com/OISF/suricata/pull/6268#discussion_r675693271)
I probably got it wrong and the idea was to have another unidirectional transaction, but representing just the response?